### ISSUE TITLE
Including images in templates?

### ISSUE THREAD

**Issue by thomthom** [lsegal/yard#1329]
I wanted to add some images to our custom template, adding them to `default/fulldoc/html/images`. But I don't know how to refer to them correctly from all the HTML pages generated.

If I prefix the paths with `/` (`/images/logo.png` ) then it will not work on local environment, only when deployed to a server environment.

Is there any way to provide images in templates that will also work in local builds?

**Comment by thomthom** [comment]
Oh... and I just realized that that directories in the templates folder isn't automatically copied.

How do one go about telling YARD which files to copy from the template?

**Comment by lsegal** [comment]
YARD doesn't copy files from templates because they are templates. Generally speaking the way to copy assets is to use `--asset`, but if you're trying to do this as part of a template, the recommended path is to actually not rely on separable assets and _embed_ your assets into templates. Basically in short, use data URLs to [embed assets like images in HTML or CSS documents](https://github.com/lsegal/yard/blob/master/templates/default/fulldoc/html/css/style.css#L351).

YARD templates _can_ manually copy assets using the `asset` method (see [example](https://github.com/lsegal/yard/blob/master/templates/default/fulldoc/html/setup.rb#L121) here), but doing so is pretty fraught because those asset paths change(*) depending on if you're accessing via static generation, YARD server, or another tool that your template cannot control for.

TLDR: embed image assets into templates as much as possible to avoid issues with cross-referencing files across varying environments.

(*) for example: HTML files do not route with the .html extension when using server, so linking to them as such would not work-- we handle this specially with `link_*` helper methods, but you may end up in a similar situation with image or other asset routing. Rubydoc.info, for instance, modifies routes to add in global css documents to improve caching and provide extra stylesheets, and custom templates _could_ break this by putting dependent files in locations apart from the related stylesheets.

**Comment by thomthom** [comment]
Thanks for replying.

One of the assets I wanted to add was a favicon, which wasn't embeddable.

I did in fact end up using `asset` to perform the copy.
https://github.com/SketchUp/sketchup-yard-template/commit/fa19368c40e2525c38d67c1a4bc800e408022183#diff-3e3556ffe567b134a8368986065d1974

I meant to post back here to mention that to see if that was a sane way to do it.

In our case the template is a custom template specifically for us and not general usage. So the intended usage is only our own and `assert` worked fine for that.
